AP_RangeFinder: add support for RDS02UF radar driver on seria

zebulon-86 opened the pull request:
## Summary

Resubmit pull request，relevant : https://github.com/ArduPilot/ardupilot/pull/22385
Adding new features
AP_RangeFinder:add support for RDS02UF radar driver on serial。

## Documentation

Product URL：[Microbrain Intelligent](https://www.microbrain.com.cn/en/ProductCenter/list.aspx?categoryid=132)
Related Manuals:[【Benewake(北醒)】毫米波雷达在无人机上避障资料汇总](https://blog.csdn.net/zoran_wu/article/details/126750128)
Here are the dumped Manuals:[RDS02U_Uart_Manual](https://github.com/zebulon-86/Document/blob/master/MicrobrainRadar/RDS02U_Uart_Manual.md)
For PDF:[RDS02U](https://github.com/zebulon-86/Document/tree/master/MicrobrainRadar/pdf)

## Change

Some tweaks to the code related to #22385

[libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h](https://github.com/ArduPilot/ardupilot/pull/22385/files#diff-bf4ef8c6a1897e12756a77e6efd92c249789ccd5788f54610ad9a6d1e6140622)
```c
#if RDS02_USE_CRC
    const uint8_t crc8_table[256] = {
    0x93,0x98,0xE4,0x46,0xEB,0xBA,0x04,0x4C,
```
The CRC table takes up too much flash space,so added the define of whether to enable CRC, which is not enabled by default, and only use the frame header and frame tail to judge the integrity of the frame. It is not a big problem not to do CRC when the serial communication is stable. Users can also choose to enable CRC according to the hardware situation.

magicrub commented:
Why not use existing crc functions?

muramura commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,225 @@
+
+                }
+            }
+			parser_index = 0;
```
ardupilot indentation is whitespace 4.

zebulon-86 replied:
I didn't notice that tabs are used here, I will make a modification, thank you for pointing out the problem

rmackay9 commented:
Hi @zebulon-86,

It looks like this is failing CI because of a problem with a definition.

[ 469/1015] Compiling libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.cpp
../../libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:155:13: error: "RDS02_USE_CRC" is not defined, evaluates to 0 [-Werror=undef]
  155 |         #if RDS02_USE_CRC
      |             ^~~~~~~~~~~~~

Hwurzburg reviewed (changes requested):
this has no metadata update to the rf type param

zebulon-86 commented:
> Hi @zebulon-86,
> 
> It looks like this is failing CI because of a problem with a definition.
> 
> [ 469/1015] Compiling libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.cpp ../../libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:155:13: error: "RDS02_USE_CRC" is not defined, evaluates to 0 [-Werror=undef] 155 | #if RDS02_USE_CRC | ^~~~~~~~~~~~~

I added in build_options.py
```
Feature('Rangefinder', 'RANGEFINDER_RDS02UF', 'AP_RANGEFINDER_RDS02UF_ENABLED', "Enable Rangefinder - RDS02UF", 0, "RANGEFINDER"), # NOQA: E501
```
Added `#if AP_RANGEFINDER_RDS02UF_ENABLED`  in AP_RangeFinder_RDS02UF.cpp, now All checks have passed
@rmackay9

zebulon-86 commented:
> this has no metadata update to the rf type param

@Hwurzburg 
Sorry, I don't quite understand which one this refers to specifically. Is it the following definition?
```
MAV_DISTANCE_SENSOR _get_mav_distance_sensor_type() const override {
    return MAV_DISTANCE_SENSOR_RADAR;
}
```

rmackay9 commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,227 @@
+        case RDS02UF_PARSE_STATE::STATE7_LENGTH_H: // lengh_high
+        {
+            uint8_t read_len = data << 8 | parser_buffer[parser_index-1];
+            if (read_len == RDS02UF_DATA_LEN)	// rds02uf data length is always 10
```
nitpick: bracket should be on the earlier line

rmackay9 commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:
```diff
@@ -0,0 +1,131 @@
+    bool decode(uint8_t c);
+
+#if RDS02_USE_CRC
+    const uint8_t crc8_table[256] = {
```
This should be moved to AP_Math's crc file

zebulon-86 replied:
I submitted a commit and moved the crc table to AP_Math/crc.h

```
#define RDS02_USE_CRC 0
#if RDS02_USE_CRC
    const uint8_t rds02_crc8_table[256] = {
```

Because this table may not be applicable elsewhere, RDS02_USE_CRC is added, and this table is not enabled by default

zebulon-86 replied:
I don't know if it is appropriate to add this crc8 table to AP_Math, because this table is a fixed value defined by the manufacturer and not applicable to other devices. Because there are frame header and frame tail restrictions, there will be no problem in short-distance serial communication if the crc check is not applicable, and this check table can also be deleted directly.

rmackay9 commented on libraries/SITL/SIM_RF_RDS02UF.cpp:
```diff
@@ -0,0 +1,124 @@
+using namespace SITL;
+
+
+    const uint8_t crc8_table[256] = {
```
we should remove this and use the version in AP_Math once it has been moved.

zebulon-86 replied:
Already moved

rmackay9 commented on Tools/autotest/arducopter.py:
```diff
@@ -7932,6 +7932,7 @@ def RangeFinderDrivers(self):
             ("benewake_tf03", 27),
             ("gyus42v2", 31),
             ("teraranger_serial", 35),
+            ("rds02uf", 37),
```
This should be moved to a separate commit starting with "Tools:".

Hwurzburg commented:
add metadata in https://github.com/ArduPilot/ardupilot/blob/8653bf7349fcd98e7a38db4af9a73bda83e8e186/libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp#L14

rmackay9 commented on libraries/AP_HAL/SIMState.cpp:
```diff
@@ -163,6 +163,9 @@ void SIMState::fdm_input_local(void)
     if (leddarone != nullptr) {
         leddarone->update(sitl_model->rangefinder_range());
     }
+    if (rds02uf != nullptr) {
```
These AP_HAL_SIMState changes should be moved to a separate commit prefixed with "AP_HAL:"

rmackay9 commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,227 @@
+    uint8_t data = c;
+    bool ret = false;
+    switch (decode_state)
+    {
```
nitpick: bracket should be one line higher

rmackay9 commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,227 @@
+#if RDS02_USE_CRC
+     uint8_t crc = 0;
+     uint8_t* data = pbuf;
+     while ( len-- )
```
let's remove the white space between the brackets.

zebulon-86 replied:
Thanks for the reminder, I'm not careful enough about the format, I checked again

zebulon-86 commented:
> also need to add to extract_features.py when adding a feature to build server

I added the following code to the extract_features.py file.

```
('AP_RANGEFINDER_RDS02UF_ENABLED', r'AP_RangeFinder_RDS02UF::get_reading\b',),
```

peterbarker commented:
All of my review comments on https://github.com/ArduPilot/ardupilot/pull/21646 almost certainly still apply.

Please do not open multiple PRs, but force-push over the branch of the previous PR.

rishabsingh3003 commented:
@zebulon-86 I have rebased this PR. I have also fixed the issues mentioned above and made a few changes to the PR. The changes I have made are in a different commit.

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,175 @@
+    // read any available data from the radar
+    float sum = 0.0f;
+    uint16_t count = 0;
+    uint32_t nbytes = MAX(uart->available(), 1024U);
```
should be MIN() not MAX()

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,175 @@
+    uint32_t nbytes = MAX(uart->available(), 1024U);
+    float distance_m;
+    while (nbytes-- > 0) {
+        const int16_t c = uart->read();
```
should use the bool version of read()

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,175 @@
+            break;
+        case RDS02UF_PARSE_STATE::STATE7_LENGTH_H:{
+            // length_high
+            uint8_t read_len = data << 8 | parser_buffer[parser_index-1];
```
how do you get data<<8 into a uint8_t ?

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,175 @@
+            // length_high
+            uint8_t read_len = data << 8 | parser_buffer[parser_index-1];
+            // rds02uf data length is always 10
+            if (read_len == RDS02UF_DATA_LEN) {
```
need a check for read_len longer than buffer length of 50

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,175 @@
+			parser_index = 0;
+			decode_state = RDS02UF_PARSE_STATE::STATE0_SYNC_1;
+			break;
+        }
```
check index == sizeof(buffer), and reset at this point, so we don't put any more bytes in

tridge commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:
```diff
@@ -0,0 +1,114 @@
+#include "AP_RangeFinder.h"
+#include "AP_RangeFinder_Backend_Serial.h"
+
+#define RDS02_BUFFER_SIZE           50
```
move most defines to cpp file

rishabsingh3003 commented:
From Dev Call discussion - Redo the implementation with structure

tridge commented:
@peterbarker will review parser and merge

rishabsingh3003 commented:
Added "memchr"

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,127 @@
+#define RDS02_HEAD2                 0x55
+#define RDS02_END                   0xAA
+#define RDS02UF_PRE_DATA_LEN        6
+#define RDS02_TARGET_FC_INDEX_L     8
+#define RDS02_TARGET_FC_INDEX_H     9
```
These aren't used now

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,127 @@
+void AP_RangeFinder_RDS02UF::move_header_in_buffer(uint8_t initial_offset)
+{
+    uint8_t* header_ptr = (uint8_t*)memchr(&u.parse_buffer[initial_offset], RDS02_HEAD1, body_length - initial_offset);
+    if (header_ptr != nullptr) {
```
This doesn't look right.  In the case we don't find anything we won't empty the buffer?

AP_ExternalAHRS_VectorNAV does this, for example.

rishabsingh3003 replied:
Thanks for finding this!

peterbarker reviewed (changes requested):
Please test this with a little bit of corruption, make sure it still generally works (AP_SIM_SERIALDEVICE_CORRUPTION_ENABLED).  Run under valgrind when doing this.

rishabsingh3003 commented:
rebased, and fixed the pending comments

rishabsingh3003 commented:
Also tested with, AP_SIM_SERIALDEVICE_CORRUPTION_ENABLED and valgrind. Works well.
Tested with real sensor connected to SITL as well.

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,125 @@
+        }
+    }
+
+    // reset buffer
+    body_length = 0;
```
```suggestion
    // consume packet
    move_header_in_buffer(sizeof(u.packet));
```

The current fixed size makes this NFC, but if we wanted to be able to read multiple packets at once this would be required....

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:
```diff
@@ -0,0 +1,125 @@
+    move_header_in_buffer(0);
+
+    // header byte 1 is correct.
+    if (body_length < ARRAY_SIZE(u.parse_buffer)) {
```
```suggestion
    if (body_length < sizeof(u.packet)) {
```

... this is, after all, exactly what you're interested in.

The parse buffer can be larger than the packet, allowing for multiple packets to be read in at once (and different packets to be parsed out of the same buffer, of course)

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:
```diff
@@ -0,0 +1,102 @@
+
+    union RDS02UF_Union {
+        uint8_t parse_buffer[21];
+        struct RDS02UFPacket packet;
```
Technically we really should mark this as `PACKED`.  Since all members are 8-byte quantities it doesn't really matter.

peterbarker commented:
Merged, thanks all!

Hwurzburg commented:
@zebulon-86 I can find no vendor offering this device and Benewake does not list on its web page.....I cant see any reason to have a wiki entry for this device since its not available

rishabsingh3003 commented:
@Hwurzburg I believe Benewake plans to sell these. We have a chat with them where we could ask

Hwurzburg commented:
@zebulon-86 as far as I can tell this product does not exist and cannot be obtained......no wiki page will be generated for it at this time due to this

Hwurzburg commented:
@rishabsingh3003 I see no chat in Skype or Discord for them.....if someone can give me a link as to where to obtain this device, I will create a wiki page and entry then
